rpc/types/events.go

```diff
@@ -0,0 +1,223 @@
+	// tx_log(txLog, txLog, ...)
+	// ethereum_tx(amount, ethereumTxHash, txIndex, txGasUsed, txHash, [receipient], ethereumTxFailed)
+	// tx_log(txLog, txLog, ...)
+	// ...
+	// ```
+	// If the transaction exceeds block gas limit, it only emits the first part.
+	eventFormat2
+)
+
+// ParsedTx is the tx infos parsed from events.
+type ParsedTx struct {
+	MsgIndex int
+
+	// the following fields are parsed from events
+
+	Hash common.Hash
+	// -1 means uninitialized
+	EthTxIndex int64
+	GasUsed    uint64
+	Failed     bool
+	// unparsed tx log json strings
+	RawLogs [][]byte
+}
+
+// NewParsedTx initialize a ParsedTx
+func NewParsedTx(msgIndex int) ParsedTx {
+	return ParsedTx{MsgIndex: msgIndex, EthTxIndex: -1}
+}
+
+// ParseTxLogs decode the raw logs into ethereum format.
+func (p ParsedTx) ParseTxLogs() ([]*ethtypes.Log, error) {
+	logs := make([]*evmtypes.Log, 0, len(p.RawLogs))
+	for _, raw := range p.RawLogs {
+		var log evmtypes.Log
+		if err := json.Unmarshal(raw, &log); err != nil {
+			return nil, err
+		}
+
+		logs = append(logs, &log)
+	}
+	return evmtypes.LogsToEthereum(logs), nil
+}
+
+// ParsedTxs is the tx infos parsed from eth tx events.
+type ParsedTxs struct {
+	// one item per message
+	Txs []ParsedTx
+	// map tx hash to msg index
+	TxHashes map[common.Hash]int
+}
+
+// ParseTxResult parse eth tx infos from cosmos-sdk events.
+// It supports two event formats, the formats are described in the comments of the format constants.
+func ParseTxResult(result *abci.ResponseDeliverTx) (*ParsedTxs, error) {
+	format := eventFormatUnknown
+	// the index of current ethereum_tx event in format 1 or the second part of format 2
+	eventIndex := -1
+	var txs []ParsedTx
+	txHashes := make(map[common.Hash]int)
+	for _, event := range result.Events {
+		switch event.Type {
+		case evmtypes.EventTypeEthereumTx:
+			if format == eventFormatUnknown {
+				// discover the format version by inspect the first ethereum_tx event.
+				if len(event.Attributes) > 2 {
+					format = eventFormat1
+				} else {
+					format = eventFormat2
+				}
+			}
+
+			if len(event.Attributes) == 2 {
+				// the first part of format 2
+				msgIndex := len(txs)
+				tx := NewParsedTx(msgIndex)
+				if err := fillTxAttributes(&tx, event.Attributes); err != nil {
+					return nil, err
+				}
+				txs = append(txs, tx)
+				txHashes[tx.Hash] = msgIndex
+			} else {
+				// format 1 or second part of format 2
+				eventIndex++
+				if format == eventFormat1 {
+					// append tx
+					msgIndex := len(txs)
+					tx := NewParsedTx(msgIndex)
+					if err := fillTxAttributes(&tx, event.Attributes); err != nil {
+						return nil, err
+					}
+					txs = append(txs, tx)
+					txHashes[tx.Hash] = msgIndex
+				} else {
+					// the second part of format 2, update tx fields
+					if err := fillTxAttributes(&txs[eventIndex], event.Attributes); err != nil {
+						return nil, err
+					}
+				}
+			}
```

Comment:
I feel we can reduce the complexity of these lines by splitting them into multiple smaller functions

Reply:
indeed, I just did some refactoring.